feat(cli/input-prompt): burst paste buffering with placeholders

944750720 opened the pull request:
## TLDR

<!-- Add a brief description of what this pull request changes and why and any important things for reviewers to look at -->

## Dive Deeper

<!-- more thoughts and in-depth discussion here -->

## Reviewer Test Plan

<!-- when a person reviews your code they should ideally be pulling and running that code. How would they validate your change works and if relevant what are some good classes of example prompts and ways they can exercise your changes -->

## Testing Matrix

<!-- Before submitting please validate your changes on as many of these options as possible -->

|          | 🍏  | 🪟  | 🐧  |
| -------- | --- | --- | --- |
| npm run  | ❓  | ❓  | ❓  |
| npx      | ❓  | ❓  | ❓  |
| Docker   | ❓  | ❓  | ❓  |
| Podman   | ❓  | -   | -   |
| Seatbelt | ❓  | -   | -   |

## Linked issues / bugs
- Fixed #2627 
<!--
Link to any related issues or bugs.

**If this PR fully resolves the issue, use one of the following keywords to automatically close the issue when this PR is merged:**

- Closes #<issue_number>
- Fixes #<issue_number>
- Resolves #<issue_number>

*Example: `Resolves #123`*

**If this PR is only related to an issue or is a partial fix, simply reference the issue number without a keyword:**

*Example: `This PR makes progress on #456` or `Related to #789`*
-->

944750720 commented:
/gemini review

jacob314 commented on packages/cli/src/ui/components/InputPrompt.tsx:
```diff
@@ -210,37 +249,76 @@ export const InputPrompt: React.FC<InputPromptProps> = ({
-  }, [buffer, config]);
+  }, [buffer, config, insertAtOffsetWithAutoSpaces]);
+
+  const expandPlaceholders = useCallback(
```
as a follow up, please integrate this with the syntax highlighting for the input prompt. that will make it clear these placeholders are special and have been matched. This will make the case where the placeholder is edited less confusing as well as it will switch to plain text when it doesn't match.

jacob314 commented on packages/cli/src/ui/components/InputPrompt.tsx:
```diff
@@ -264,8 +342,13 @@ export const InputPrompt: React.FC<InputPromptProps> = ({
 
-        // Ensure we never accidentally interpret paste as regular input.
-        buffer.handleInput(key);
+        const content = key.sequence || '';
```
move all of this logic into text buffer. that will work better to integrate it into the reducer.

944750720 commented:
@jacob314 I've tried to simplify and extract code to find placeholders. And renamed highlight.ts to parse.ts. Added unittest. Please take a look when convenient.

944750720 commented on packages/cli/src/ui/components/shared/text-buffer.ts:
```diff
@@ -1020,7 +1047,94 @@ function textBufferReducerLogic(
+          const newPending = [...replaced.pendingPastes];
+          for (let j = 0; j < newPending.length; j++) {
+            if (newPending[j].placeholder === hitPlaceholder.text) {
+              if (count === sameTextBefore) {
```
@jacob314  When we have **multiple identical placeholders with different content**, we need to **precisely identify which specific instance** the user deleted. For example:

```typescript
// User has pasted 3 times with same length (same placeholder text)
pendingPastes = [
  { placeholder: '[Pasted Content 100 chars]', content: 'AAA...(100 A's)' },
  { placeholder: '[Pasted Content 100 chars]', content: 'BBB...(100 B's)' },
  { placeholder: '[Pasted Content 100 chars]', content: 'CCC...(100 C's)' }
]

text = "[Pasted Content 100 chars] [Pasted Content 100 chars] [Pasted Content 100 chars]"
```

If the user backspaces to delete the **middle** placeholder:

```typescript
text = "[Pasted Content 100 chars] [Pasted Content 100 chars]"
// We should keep: [AAA, CCC] ✅
// NOT: [AAA, BBB] ❌
```
So I chose:
```typescript
// sameTextBefore: How many placeholders with the SAME text appear BEFORE the one we're deleting
const sameTextBefore = placeholderCandidates
  .slice(0, hitIndex + 1)  // Include the current one
  .filter((ph) => ph.text === hitPlaceholder.text).length - 1;  // Subtract 1 to get "before"

// Then we find the Nth pendingPaste with this placeholder text and remove it
let count = 0;
for (let j = 0; j < newPending.length; j++) {
  if (newPending[j].placeholder === hitPlaceholder.text) {
    if (count === sameTextBefore) {  // Found the Nth one
      newPending.splice(j, 1);  // Remove it
      break;
    }
    count++;
  }
}
```
